File: Depth_First_Search/1034.Coloring_A_Border/C++/1034.Coloring_A_Border.cpp
class Solution {
public:
    vector<vector<int>> colorBorder(vector<vector<int>> & grid, int row, int col, int color) 
    {
        //============================================//
        // Input type:                                //
        //   - vector<vector<int>> ref. (grid vector) //
        //   - int (row index)                        //
        //   - int (column index)                     //
        //   - int (color value)                      //
        // Return type:                               //
        //   - vector<vector<int>> (result vector)    //
        //============================================//

        /*Initialize*/
        ///// Length of rows, columns /////
        int len_rows = grid.size(), len_cols = (grid[0]).size();

        ///// Adjacent value /////
        int adj_val = (grid[row])[col];

        ///// Record grid vector /////
        vector<vector<int>> record_grid(grid.begin(), grid.end());


        /*Recursion-based DFS loop traversal*/
        dfsTraversal(record_grid, len_rows, len_cols, row, col, adj_val); //Recursion function call

        for (int rows_idx = 0; (rows_idx < len_rows); (rows_idx++)) //Rows
        {
            for (int cols_idx = 0; (cols_idx < len_cols); (cols_idx++)) //Columns
            {
                ///// Check if the current edges boundary matched conditions or not /////
                if (((rows_idx == 0) || (rows_idx == (len_rows - 1))) || ((cols_idx == 0) || (cols_idx == (len_cols - 1))))
                {
                    ///// Check if the current indexed-value matched conditions or not /////
                    if ((record_grid[rows_idx])[cols_idx] == 1001) { (grid[rows_idx])[cols_idx] = color; } //Keep updating, overwriting
                    else { ; }
                }
                else
                {
                    ///// Check if the current indexed-value matched conditions or not /////
                    if ((record_grid[rows_idx])[cols_idx] == 1001)
                    {
                        ///// Check if the current boundary values matched conditions or not /////
                        if (((record_grid[rows_idx])[(cols_idx - 1)] != 1001) || ((record_grid[(rows_idx - 1)])[cols_idx] != 1001) || ((record_grid[rows_idx])[(cols_idx + 1)] != 1001) || ((record_grid[(rows_idx + 1)])[cols_idx] != 1001))
                        {
                            (grid[rows_idx])[cols_idx] = color; //Keep updating, overwriting

                            continue;
                        }
                    }
                    else { ; }
                }

            } //Columns

        } //Rows

        return grid;
    }


    void dfsTraversal(vector<vector<int>> & record_grid, int & len_rows, int & len_cols, int rows_idx, int cols_idx, int & adj_val)
    {
        //===================================================//
        // Input type:                                       //
        //   - vector<vector<int>> ref. (record grid vector) //
        //   - int ref. (length of rows)                     //
        //   - int ref. (length of columns)                  // 
        //   - int (row index)                               //
        //   - int (column index)                            //
        //   - int ref. (adjacent value)                     //
        // Return type:                                      //
        //   - void (no return)                              //
        //===================================================//
        if (((rows_idx < 0) || (rows_idx >= len_rows)) || ((cols_idx < 0) || (cols_idx >= len_cols))) { return; } //Issue, Boundary-case handling
        if ((record_grid[rows_idx])[cols_idx] != adj_val) { return; } //Issue, Boundary-case handling

        /*Whole process, flow*/
        (record_grid[rows_idx])[cols_idx] = 1001; //Update, Overwrite

        dfsTraversal(record_grid, len_rows, len_cols, rows_idx, (cols_idx - 1), adj_val); //Recursion function call (i.e. left)
        dfsTraversal(record_grid, len_rows, len_cols, (rows_idx - 1), cols_idx, adj_val); //Recursion function call (i.e. up)
        dfsTraversal(record_grid, len_rows, len_cols, rows_idx, (cols_idx + 1), adj_val); //Recursion function call (i.e. right)
        dfsTraversal(record_grid, len_rows, len_cols, (rows_idx + 1), cols_idx, adj_val); //Recursion function call (i.e. down)
    }
};

File: Depth_First_Search/1034.Coloring_A_Border/Python/1034.Coloring_A_Border.py
class Solution(object):
    def colorBorder(self, grid, row, col, color):
        """
        :type grid: List[List[int]]
        :type row: int
        :type col: int
        :type color: int
        :rtype: List[List[int]]
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#

        ############
        #Initialize
        ##### Length of rows, columns #####
        len_rows, len_cols = len(grid), len(grid[0])

        ##### Adjacent value #####
        adj_val = (grid[row])[col]

        ##### Record grid array #####
        record_grid = deepcopy(grid)


        ####################################
        #Recursion-based DFS loop traversal
        self.dfsTraversal(record_grid, len_rows, len_cols, row, col, adj_val) #Recursion function call

        for rows_idx in range(len_rows):
            for cols_idx in range(len_cols):

                ##### Check if the current edge boundary matched conditions or not #####
                if (((not rows_idx) or (rows_idx == (len_rows - 1))) or ((not cols_idx) or (cols_idx == (len_cols - 1)))):

                    ##### Check if the current indexed-value matched conditions or not #####
                    if ((record_grid[rows_idx])[cols_idx] == 1001):
                        (grid[rows_idx])[cols_idx] = color #Keep updating/overwriting
                    else:
                        pass

                else:
                    ##### Check if the current indexed-value matched conditions or not #####
                    if ((record_grid[rows_idx])[cols_idx] == 1001):

                        ##### Check if the current boundary value matched conditions or not #####
                        if (((record_grid[rows_idx])[(cols_idx - 1)] != 1001) or ((record_grid[(rows_idx - 1)])[cols_idx] != 1001) or ((record_grid[rows_idx])[(cols_idx + 1)] != 1001) or ((record_grid[(rows_idx + 1)])[cols_idx] != 1001)):
                            (grid[rows_idx])[cols_idx] = color #Keep updating/overwriting
                        else:
                            pass
                    else:
                        pass
        
        return grid


    def dfsTraversal(self, record_grid, len_rows, len_cols, rows_idx, cols_idx, adj_val):
        """
        :type record_grid: List[List[int]]
        :type len_rows: int
        :type len_cols: int
        :type rows_idx: int
        :type cols_idx: int
        :type adj_val: int
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (((rows_idx < 0) or (rows_idx >= len_rows)) or ((cols_idx < 0) or (cols_idx >= len_cols))): #Issue/Boundary-case handling
            return

        if ((record_grid[rows_idx])[cols_idx] != adj_val): #Issue/Boundary-case handling
            return


        ####################
        #Whole process/flow
        (record_grid[rows_idx])[cols_idx] = 1001 #Keep updating/overwriting

        self.dfsTraversal(record_grid, len_rows, len_cols, rows_idx, (cols_idx - 1), adj_val) #Recursion function call (i.e. left)
        self.dfsTraversal(record_grid, len_rows, len_cols, (rows_idx - 1), cols_idx, adj_val) #Recursion function call (i.e. up)
        self.dfsTraversal(record_grid, len_rows, len_cols, rows_idx, (cols_idx + 1), adj_val) #Recursion function call (i.e. right)
        self.dfsTraversal(record_grid, len_rows, len_cols, (rows_idx + 1), cols_idx, adj_val) #Recursion function call (i.e. down)
